google/oauth2/credentials.py

```diff
@@ -116,13 +124,18 @@ def __init__(
         self._client_secret = client_secret
         self._quota_project_id = quota_project_id
         self._rapt_token = rapt_token
+        self.refresh_handler = refresh_handler
 
     def __getstate__(self):
         """A __getstate__ method must exist for the __setstate__ to be called
         This is identical to the default implementation.
         See https://docs.python.org/3.7/library/pickle.html#object.__setstate__
         """
-        return self.__dict__
+        state_dict = self.__dict__.copy()
+        # Remove _refresh_handler function as functions can't be pickled.
+        # The refresh_handler setter should be used to repopulate this.
+        del state_dict["_refresh_handler"]
+        return state_dict
```

Comment:
I'm not sure why you state, "functions can't be pickled":

```python
>>> import pickle
>>> def foo(): pass
... 
>>> bar = {"foo": foo}
>>> dumped = pickle.dumps(bar)
>>> pickle.loads(dumped)
{'foo': <function foo at 0x7fc3a19c6598>}
>>> dumped
b'\x80\x03}q\x00X\x03\x00\x00\x00fooq\x01c__main__\nfoo\nq\x02s.'
```

Some kinds of on-the-fly constructed functions (`lambda`, `functools.partial` instances) cannot be pickled, because they need to be importable.  See [what can be pickled](https://docs.python.org/3/library/pickle.html#what-can-be-pickled-and-unpickled) in the Python docs.

Reply:
Thanks for the reference. 
I ran into various errors when pickling a mock function and a local function:
```
    def test_pickle_and_unpickle_with_refresh_handler(self):
        expected_expiry = _helpers.utcnow() + datetime.timedelta(seconds=2800)
        # refresh_handler = mock.Mock(return_value=("TOKEN", expected_expiry))
        def refresh_handler():
            return ("TOKEN", expected_expiry)
    
        creds = credentials.Credentials(
            token=None,
            refresh_token=None,
            token_uri=None,
            client_id=None,
            client_secret=None,
            rapt_token=None,
            refresh_handler=refresh_handler,
        )
>       unpickled = pickle.loads(pickle.dumps(creds))
E       AttributeError: Can't pickle local object 'TestCredentials.test_pickle_and_unpickle_with_refresh_handler.<locals>.refresh_handler'
```
As my comment is not accurate, I have updated it to be more accurate.